[SPARK-33482][SQL] Fix FileScan canonicalization

peter-toth opened the pull request:
### What changes were proposed in this pull request?

This PR adds canonicalization to `FileScan.partitionFilters` and `FileScan.dataFilters` in `BatchScanExec` nodes.

### Why are the changes needed?

Partition filters and data filters added to `FileScan` (in https://github.com/apache/spark/pull/27112 and https://github.com/apache/spark/pull/27157) caused that canonicalized form of some `BatchScanExec` nodes don't match and this prevents some reuse possibilities.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Added new UT.

peter-toth committed:
[SPARK-33482][SQL] Fix FileScan canonicalization

peter-toth commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala:
```diff
@@ -86,7 +86,7 @@ trait FileScan extends Scan
   override def equals(obj: Any): Boolean = obj match {
     case f: FileScan =>
-      fileIndex == f.fileIndex && readSchema == f.readSchema
+      fileIndex == f.fileIndex && readSchema == f.readSchema &&
```
This change is not related to this PR, but it looks like a `&&` is missing from here.

c21 replied:
Nice catch. It seems that https://github.com/apache/spark/pull/27112 introduced this. 
 cc @dongjoon-hyun if we want to backport this to `3.1` and `3.0`.

maropu replied:
Oh... nice catch. cc: @gengliangwang , too.

peter-toth replied:
Credit goes to @bersprockets for catching this in SPARK-33482.

HyukjinKwon replied:
wow ..

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40602/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40602/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/40602/

c21 commented on sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:
```diff
@@ -4065,6 +4066,33 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
+                |JOIN t AS t2 ON t2.id = t1.id
+                |JOIN t AS t3 ON t3.id = t2.id
+                |""".stripMargin)
+            df.collect()
+            df.explain()
```
do we really need these two statements?

peter-toth replied:
No. I left `explain()` there accidentally, removed in: https://github.com/apache/spark/pull/31820/commits/c0fb9b2cdca15e06ff2f708b7046a44cfc965da1

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/136018/

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:
```diff
@@ -4065,6 +4066,33 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
+  test("SPARK-33482: Fix FileScan canonicalization") {
+    Seq(true, false).foreach { aqe =>
+      withSQLConf(SQLConf.USE_V1_SOURCE_LIST.key -> "",
+        SQLConf.ADAPTIVE_EXECUTION_ENABLED.key -> aqe.toString) {
```
We need to set this config for this test purpose?

peter-toth replied:
Well, reuse exchange code is very different in AQE/non-AQE paths, but I think you are right as we just need to test the canonicalization fix so I've removed this in https://github.com/apache/spark/pull/31820/commits/c0fb9b2cdca15e06ff2f708b7046a44cfc965da1

peter-toth committed:
address comments

SparkQA commented:
**[Test build #136026 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/136026/testReport)** for PR 31820 at commit [`c0fb9b2`](https://github.com/apache/spark/commit/c0fb9b2cdca15e06ff2f708b7046a44cfc965da1).

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/40610/

SparkQA commented:
**[Test build #136026 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/136026/testReport)** for PR 31820 at commit [`c0fb9b2`](https://github.com/apache/spark/commit/c0fb9b2cdca15e06ff2f708b7046a44cfc965da1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/136026/

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:
```diff
@@ -49,6 +49,13 @@ case class BatchScanExec(
-    this.copy(output = output.map(QueryPlan.normalizeExpressions(_, output)))
+    val canonicalizedScan = scan match {
+      case s: FileScan =>
+        s.withFilters(QueryPlan.normalizePredicates(s.partitionFilters, output),
+          QueryPlan.normalizePredicates(s.dataFilters, output))
```
```suggestion
        s.withFilters(
          QueryPlan.normalizePredicates(s.partitionFilters, output),
          QueryPlan.normalizePredicates(s.dataFilters, output))
```

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:
```diff
@@ -49,6 +49,13 @@ case class BatchScanExec(
+          QueryPlan.normalizePredicates(s.dataFilters, output))
+      case _ => scan
+    }
+    this.copy(output = output.map(QueryPlan.normalizeExpressions(_, output)),
+      scan = canonicalizedScan)
```
```suggestion
    this.copy(
      output = output.map(QueryPlan.normalizeExpressions(_, output)),
      scan = canonicalizedScan)
```

c21 commented on sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:
```diff
@@ -4065,6 +4066,29 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
+              |JOIN t AS t2 ON t2.id = t1.id
+              |JOIN t AS t3 ON t3.id = t2.id
+              |""".stripMargin)
+          df.collect()
```
do we really need `df.collect()` here? shouldn't `AdaptiveSparkPlanHelper.collect()` below take care of going through query plan properly with AQE being enabled?

peter-toth replied:
Yes, we do need to run the query first and then check the plan as this is an AQE compatible query where `ReusedExchangeExec` nodes are inserted during execution.

c21 reviewed (approved):
LGTM with minor test comment.

peter-toth committed:
Update sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala

Co-authored-by: Hyukjin Kwon <gurwls223@gmail.com>

peter-toth committed:
Update sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala

Co-authored-by: Hyukjin Kwon <gurwls223@gmail.com>

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/40622/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/136039/

dongjoon-hyun reviewed (commented):
This is a nice patch, @peter-toth . For a better traceability, please proceed `FileScan` change in a separate PR. It looks worth to have a new JIRA because it looks like a correctness issue. And, if possible, with a separate test case focus on `FileScan.equals`.

dongjoon-hyun commented:
Also, cc @cloud-fan .

peter-toth committed:
revert FileScan.equals change from this PR

peter-toth commented:
> This is a nice patch, @peter-toth . For a better traceability, please proceed `FileScan` change in a separate PR. It looks worth to have a new JIRA because it looks like a correctness issue. And, if possible, with a separate test case focus on `FileScan.equals`.

All right, reverted in https://github.com/apache/spark/pull/31820/commits/483686d98e887c278974c0ef89bf32cbb22b580d and extracted to https://github.com/apache/spark/pull/31848

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/136105/

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/BatchScanExec.scala:
```diff
@@ -49,6 +49,15 @@ case class BatchScanExec(
+      case s: FileScan =>
+        s.withFilters(
+          QueryPlan.normalizePredicates(s.partitionFilters, output),
+          QueryPlan.normalizePredicates(s.dataFilters, output))
```
This works, but is a bit hacky as it doesn't apply to all the `Scan` implementations.

I think we should add doc in the `Scan` interface to explain how the `hashCode/equals` should be implemented.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40687/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40687/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/40687/

peter-toth commented:
Reviewers, I've moved the e2e test to https://github.com/apache/spark/pull/31848 and fixed the issue there by changing `FileScan.equals()`.

maropu commented:
We can close this now?
